Improve dlna_dmr code quality

chishm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The DLNA-DMR integration migrated to configuration via the UI. Configuring DLNA-DMR via YAML configuration has been deprecated and will be removed in a future Home Assistant release.

Your existing YAML configuration is imported on upgrade to this release. Please check the integrations configuration page for imported DLNA-DMR configuration. Once it has been migrated it can be safely removed from your YAML configuration.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a follow up PR to https://github.com/home-assistant/core/pull/55267, to address @MartinHjelmare's [code review](https://github.com/home-assistant/core/pull/55267#pullrequestreview-768393179).

It is mostly under-the-hood improvements to code quality with one exception -- the import flow for configuration will be changed to require user interaction in the event that a device is not available to perform an automatic config entry migration. Hence the new blurb for breaking changes above.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chishm committed:
Listen for config updates from DlnaDmrEntity.async_added_to_hass

Use `Entity.async_on_remove` for dealing with callback cancellation,
instead of re-inventing the wheel with `_remove_ssdp_callbacks`.

chishm committed:
Use async_write_ha_state within async methods

chishm committed:
Import YAML config from async_setup_platform

chishm commented:
From https://github.com/home-assistant/core/pull/55267#discussion_r719814054 :
I use `del <param>` in callback functions to indicate that `<param>` is unused, and to [satisfy PyLint](http://pylint-messages.wikidot.com/messages:w0613). I realise that HA suppresses the `unused-argument` warning, but it's a habit I've gotten into from other code bases and thought it was good to keep it up.

MartinHjelmare commented:
We don't delete local variables like that currently. Please remove those so we stick to our current common practice. Thanks!

chishm committed:
Import flow prompts user when device is uncontactable during migration

When config flow is able to contact a device, or when it has information
from SSDP, it will create config entries without error. If the device is
uncontactable at this point then it will appear as unavailable in HA
until it is turned on again.

When import flow cannot migrate an entry because it needs to contact the
device and can't, it will notify the user with a config flow form.

chishm committed:
Don't del unused parameters, HA pylint doesn't care

chishm commented:
Hi @MartinHjelmare, I've finished the changes requested.

chishm committed:
Remove unused imports from tests

MartinHjelmare commented on homeassistant/components/dlna_dmr/config_flow.py:
```diff
@@ -93,133 +95,182 @@ async def async_step_user(self, user_input: FlowInput = None) -> FlowResult:
+                    "Aborting SSDP setup because migration for %s is in progress",
+                    self._location,
+                )
+                raise data_entry_flow.AbortFlow("already_in_progress")
```
We're inside a step method so we can use `self.async_abort`.

StevenLooman replied:
Won't `self.async_set_unique_id()` already abort the flow before this can be reached?

MartinHjelmare replied:
In the import flow we set the unique_id to the location while here we set the unique_id to the udn, so the `async_set_unique_id` call won't abort automatically here.

chishm replied:
Indeed. The UDN isn't known early in the import flow so the location is used instead to identify and deconflict running import flows.

MartinHjelmare commented on tests/components/dlna_dmr/test_config_flow.py:
```diff
@@ -515,28 +447,40 @@ async def test_ssdp_flow_success(hass: HomeAssistant) -> None:
+
+    # Remove the device to clean up all resources, completing its life cycle
+    entry_id = result["result"].entry_id
+    assert await hass.config_entries.async_remove(entry_id) == {
```
We don't need to remove the config entry in the config flow tests. If we want to test that it's normally part of the `test_init.p` module tests.

chishm replied:
In this case it's to satisfy the post-conditions in the `domain_data_mock` and `dmr_device_mock` fixtures. The post-conditions check that all acquired resources have been released.

I'd prefer to keep the post-conditions and the clean up in, as it has already managed to find bugs for me while I was writing the `DlnaDmrEntity` code. I can put the clean up into a helper function to make it more obvious what's going on, though.

StevenLooman commented on homeassistant/components/dlna_dmr/__init__.py:
```diff
@@ -3,39 +3,13 @@
-    for entry_config in config[MEDIA_PLAYER_DOMAIN]:
-        if entry_config.get(CONF_PLATFORM) != DOMAIN:
-            continue
-        LOGGER.warning(
```
What is normally the deprecation time for configuration changes, or rather showing warnings about it? If this is merged its only one month, since #55267, that the `dlna_dmr` component was upgraded.

MartinHjelmare replied:
We're moving the warning to the media_player platform.

MartinHjelmare replied:
The depreciation time is normally two releases (months).

chishm committed:
Abort config flow at earliest opportunity

chishm committed:
Return async_abort instead of raising AbortFlow

chishm committed:
Consolidate config entry test cleanup into a single function

chishm committed:
Merge branch 'dev' into dlna_dmr-pr-fixes

MartinHjelmare commented on tests/components/dlna_dmr/test_config_flow.py:
```diff
@@ -61,6 +61,27 @@
+    # Wait for platform to finish setup
+    await hass.async_block_till_done()
+
+    # Remove the config entry and device to clean up all resources
```
The tests won't always want to test a remove of the config entry. If we have to remove the config entry manually in the test the clean up check isn't checking the code under test but how we wrote the test. I think it's better to have this check applied selectively when we want to check that the code under test cleans up as it should.

chishm replied:
I've refactored the tests to suit this style now. Resources checks for config entries are done in `test_init.py`. Resource checks for the `DlnaDmrEntity` are done in `test_media_player.py` fixtures.

chishm committed:
Merge branch 'dev' into dlna_dmr-pr-fixes

chishm committed:
fixup! Consolidate config entry test cleanup into a single function

Revert "Consolidate config entry test cleanup into a single function"

This reverts commit 8220da7263e346a37c3e1a219dc374c4a43c7df5.

chishm committed:
Check resource acquisition/release in specific tests

chishm committed:
fixup! Check resource acquisition/release in specific tests

chishm commented:
I think this is resolves all the concerns raised now.

Before it is merged I'd like to do a rebase and force push, to clean up the git history (there's a few fixups in there).

chishm committed:
Remove unused network dependency from manifest

chishm commented:
The manifest change is to fix a minor metadata issue highlighted by emontnemery in https://github.com/home-assistant/core/pull/57059#discussion_r721973627

MartinHjelmare commented:
Please don't squash or amend commits after review has started to make it easier for readers to track changes. We'll squash when merging. Thanks!

chishm commented:
> Please don't squash or amend commits after review has started to make it easier for readers to track changes. We'll squash when merging. Thanks!

Indeed, hence I haven't :)

I'd just like to do one rebase just before merge, to clean out the fixup and revert commits. Alternatively, whoever does the merge could remove any line that starts with `fixup!` from the merge commit message.

MartinHjelmare commented on tests/components/dlna_dmr/test_media_player.py:
```diff
@@ -74,21 +77,47 @@ async def mock_entity_id(
     yield entity_id
 
     # Unload config entry to clean up
     assert await hass.config_entries.async_remove(config_entry_mock.entry_id) == {
```
Why do we need to clean up here? We now test this in `test_resource_lifecycle` so we don't need to test it again here if it's the same resources used.

MartinHjelmare replied:
The entity should not be able to acquire resources outside the lifetime of the config entry. If the config entry is removed all connected entities should also be removed.

chishm replied:
You are correct, I misspoke.

The `DlnaDmrEntity` can acquire and release resources on its own, *inside* the lifetime of the config entry, but more often than just the setup and removal of the config entry. Specifically, each time it connects to the underlying device it will acquire certain resources, and each time it disconnects it should release those resources. Connection and disconnection can happen multiple times within the lifetime of the config entry, but when the config entry is removed then the device should be disconnected (if it was connected at the time) and all resources released.

MartinHjelmare replied:
Right. And what is part of this test tests something that is not tested in `test_resource_lifecycle`?

MartinHjelmare replied:
Looks like bith the ssdp callbacks and the polling causes connection and disconnection events. Ok.

MartinHjelmare commented:
No, please don't squash or remove any commits.

MartinHjelmare reviewed (commented):
There's one remaining `schedule_update_ha_state` call on line 443. Will that run in async or sync context?

chishm committed:
_on_event runs in async context

MartinHjelmare commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -440,7 +440,7 @@ def _on_event(
             # Indicates a failure to resubscribe, check if device is still available
             self.check_available = True
-        self.schedule_update_ha_state()
+        self.async_schedule_update_ha_state()
```
```suggestion
        self.async_write_ha_state()
```

chishm committed:
Call async_write_ha_state directly (not via shedule_update)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Thanks!

chishm commented:
Would it be possible to get this cherry-picked onto the 2021.10 release? I think the async state update could help to avoid potential bugs, and the import config flow is a bit more robust and user friendly. 


-------- Original Message --------
From: Martin Hjelmare ***@***.***>
Sent: 8 October 2021 7:14:12 am AEDT
To: home-assistant/core ***@***.***>
Cc: Michael Chisholm ***@***.***>, Author ***@***.***>
Subject: Re: [home-assistant/core] Improve dlna_dmr code quality (#56886)

Merged #56886 into dev.

-- 
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub:
https://github.com/home-assistant/core/pull/56886#event-5430836394

MartinHjelmare commented:
If we consider it a new breaking change we can't release this change now. If we think it's a fix to the import flow, it could be ok.
